sherpa/astro/data.py

```diff
@@ -3563,15 +3563,19 @@ def notice(self, lo=None, hi=None, ignore=False, bkg_id=None):
             return
 
         if self.units == 'energy':
-            umin = elo[0]
-            umax = ehi[-1]
+            # It's not guaranteed that channels are in increasing energy
+            # but we can certainly assume that they are monotonic.
+            # Thus, the min of the first and last entry is the minimum
+            # energy of the elo array
+            umin = min(elo[0], elo[-1])
+            umax = max(ehi[0], ehi[-1])
         elif self.units == 'wavelength':
-            umin = self._hc / ehi[-1]
-            umax = self._hc / elo[0]
+            umin = min(self._hc / ehi[[0, -1]])
+            umax = max(self._hc / elo[[0, -1]])
         else:
             # assume channel units
-            umin = self.channel[0]
-            umax = self.channel[-1]
+            umin = min(self.channel[0], self.channel[-1])
```

Comment:
One of the things I'd like to see (not in this PR) is for `DataPHA.__init__` and it's attributes to make a numbre of sanity checks - e.g. that `.channel` and `.counts` have the same size and matches the DATCHANS value [if we store it?], that `.channel` is in increasing order, ... - so that we can more-safely access this related information.

Coming back to this change - the OGIP standard has channel stored in increasing order, so I think this check should not be needed. I'd be interested to know if the xmm data doesn't match this.

Reply:
Channel is indeed in increasing order, but OGIP does not require channel to be proportional to energy. After all, the whole concept of "channel" is needed precisely because the channel may not have a simple (and known) mapping to energy. In the XMM/RGS RMFs, the channel number is increasing, but not the energy. Instead is is increasing with wavelength (not shown in the EBOUNDS extension, but we know how to convert energy to wavelength).
```
(ciao) [Hanss-MacBook-Air:sherpa-test-data/sherpatest/xmmrgs] hamogu% dmlist "P0112880201R1S004RSPMAT1003.FTZ" blocks 
 
--------------------------------------------------------------------------------
Dataset: P0112880201R1S004RSPMAT1003.FTZ
--------------------------------------------------------------------------------
 
     Block Name                          Type         Dimensions
--------------------------------------------------------------------------------
Block    1: PRIMARY                        Null        
Block    2: MATRIX                         Table         6 cols x 4000     rows
Block    3: EBOUNDS                        Table         3 cols x 3600     rows
(ciao) [Hanss-MacBook-Air:sherpa-test-data/sherpatest/xmmrgs] hamogu% dmlist "P0112880201R1S004RSPMAT1003.FTZ[EBOUNDS]" data
 
--------------------------------------------------------------------------------
Data for Table Block EBOUNDS
--------------------------------------------------------------------------------
 
ROW    CHANNEL    E_MIN                E_MAX
 
     1          1         3.0918762684         3.0996062756
     2          2         3.0841853619         3.0918762684
     3          3         3.0765318871         3.0841853619
     4          4         3.0689170361         3.0765318871
     5          5         3.0613391399         3.0689170361
     6          6         3.0537991524         3.0613391399
     7          7         3.0462958813         3.0537991524
     8          8         3.0388295650         3.0462958813
     9          9         3.0313994884         3.0388295650
    10         10         3.0240061283         3.0313994884
    11         11         3.0166482925         3.0240061283
```
In fact, for grating spectra, "wavelength" is the natural coordinate. It's linear from the origin, it's the quantity we usually care about. However, Chandra orders in energy, even for grating spectra:
```
(ciao) [Hanss-MacBook-Air:sherpa/sherpa-test-data/sherpatest] hamogu% dmlist "3c120_heg_-1.rmf.gz[EBOUNDS]" data
 
--------------------------------------------------------------------------------
Data for Table Block EBOUNDS
--------------------------------------------------------------------------------
 
ROW    CHANNEL    E_MIN                E_MAX
 
     1          1     0.57720756530762     0.57727473974228
     2          2     0.57727473974228     0.57734197378159
     3          3     0.57734197378159     0.57740914821625
     4          4     0.57740914821625     0.57747644186020
     5          5     0.57747644186020     0.57754367589951
     6          6     0.57754367589951     0.57761090993881
     7          7     0.57761090993881     0.57767820358276
     8          8     0.57767820358276     0.57774549722672
     9          9     0.57774549722672     0.57781279087067
    10         10     0.57781279087067     0.57788014411926
    11         11     0.57788014411926     0.57794749736786
```
I think it is a moot point for us why exactly XMM chose to do it that way - in my opinion *not* supporting XMM is not an option. 

Note that the XMM data is sane, in the sense that the channel number *is* increasing, the number of columns is correct etc. It just does not follow the implicit assumption build into #1127.

Reply:
On the other hand, the MATRIX extension is ordered in **increasing** energy, but OGIP does not say that the row number in the MATRIX extension has to correspond to the order in the EBOUNDS extentions. To a certain degree, the ordering in MATRIX is arbitrary, since each row says with channels (through F_CHAN and N_CHAN) are mapped to which (through the MATRIX). Now, I can't say that I *understand* why they did it this way or that I agree that it is a good choice, (it certainly caused me enough confusion to change the title of this PR twice), but have to deal with it - which is waht this PR is about.